plugin/summernote-ext-hint.js

```diff
@@ -0,0 +1,241 @@
+      i = i || 0;
+
+      if (i < 0) {
+        i = $popover.children().length - 1;
+      }
+
+      $popover.children().removeClass('active');
+      var $activeItem = $popover.children().eq(i);
+      $activeItem.addClass('active');
+
+      this.scrollTo($activeItem);
+    },
+
+    replace: function ($popover) {
+      var wordRange = $popover.data('wordRange');
+      var $activeItem = $popover.find('.active');
+      var content = this.content($activeItem.html(), $activeItem.data('keyword'));
+
+      if (typeof content === 'string') {
+        content = document.createTextNode(content);
+      }
+
+      wordRange.insertNode(content);
+      range.createFromNode(content).collapse().select();
+    },
+
+    loadEmoji: function (data) {
+      this.emoji = Object.keys(data);
+      this.emojiLink = data;
+    },
+
+    /**
+     * @async
+     * @param {String} keyword
+     * @param {Function} callback
+     * @return {Object}
+     */
+    searchKeyword: function (keyword, callback) {
+      var triggerChar = keyword.charAt(0);
+
+      if (triggerChar === ':' && keyword.length > 1) {
+        var trigger = keyword.toLowerCase().replace(':', '');
+        callback({
+          type: 'emoji',
+          list: $.grep(this.emoji, function (item) {
+            return item.indexOf(trigger) === 0;
+          })
+        });
+      } else {
+        callback();
+      }
+    },
+
+    /**
+     * create items
+     *
+     * @param {Object} searchResult
+     * @param {String} searchResult.type
+     * @param {String[]} searchResult.list
+     * @return {jQuery[]}
+     */
+    createItems: function (searchResult) {
+      var items = [];
+      var list = searchResult.list;
+
+      for (var i = 0, len = list.length; i < len; i++) {
+        var $item = $('<a class="list-group-item"></a>');
+        $item.append(this.createItem(list[i]));
+        $item.data('keyword', list[i]);
+        items.push($item);
+      }
+
+      if (items.length) {
+        items[0].addClass('active');
+      }
+
+      return items;
+    },
+
+    /**
+     * create list item template
+     *
+     * @param {Object} item
+     * @returns {String}
+     */
+    createItem: function (item) {
+      var content = this.emojiLink[item];
+      return '<img src="' + content + '" width="20" /> :' + item + ':';
+    },
+
+    /**
+     * create inserted content to add in summernote
+     *
+     * @param {String} html
+     * @param {String} keyword
+     * @return {Node|String}
+     */
+    content: function (html, item) {
+      var url = this.emojiLink[item];
+
+      if (url) {
+        var $img = $('<img />').attr('src', url).css({
+          width : 20
+        });
+        return $img[0];
+      }
+
+      return html;
+    },
+
+    load: function () {
+      var self = this;
+      $.getJSON('https://api.github.com/emojis', function (data) {
+        self.loadEmoji(data);
+      });
+    },
+
+    init: function (layoutInfo) {
+      var self = this;
```

Comment:
I thought we need a handler for return resources(for this case events) in plugin system.
Unfortunately,There is no way to detach events.